### ISSUE TITLE
[ENH] Multiple conda environments

### ISSUE THREAD

**Issue by AlbertDeFusco** [conda-incubator/conda-project#6]
The environment.yml file specification currently only supports one env. We can use multiple yaml files combined with environment specification in `conda-project.yml` to be able to maintain and extend multiple environments in a single project.

```yaml
commands:
  server:
    cmd: ...
    env_spec: default
  test:
    cmd: pytest
    env_spec: test

env_specs:  
  default:
    filename: environment.yml  # default behavior
  test:
    filename: environment-test.yml
  dev:
    filename: environment-test.yml
    dependencies:
    - notebook
    - ...
```

**Comment by jbednar** [comment]
Personally, I find it quite confusing if multiple environments are defined in the same project except for the special case of _nested_ environments. E.g. for the `test` environment above, I'd want the tests to use an environment that contains (as far as possible) precisely the same packages as in `default`, _plus_ others only needed for testing. So I wouldn't ever want `test` to be a fully independent environment the point of it would be to test what gets installed in the default environment. And for `dev` I'd always use a separate project, independent of this one if I need an independent environment. So my personal vote would be for supporting (or at least encouraging) only the nested case, but if others can outline a good case for fully independent environments, it's not that I object to having them I just don't see the need for them and find them confusing.

**Comment by mariusvniekerk** [comment]
So conda-lock has a slightly different view of these.  It treats a regular environment.yaml file as a source of specs. It can then combine a bunch of them together so that you can get incremental environments

**Comment by mcg1969** [comment]
I do like @jbednar's idea about nested environments but I would want to see that implemented with some sort of keyword; say, `extends:`.

Multiple environments would be useful for situations where a "project" involves multiple workstreams on the same data set; e.g., ingest, interactive model development, inference REST API, etc.

**Comment by jlstevens** [comment]
For what it is worth, my experience is that extending environments is common while having entirely orthogonal environments would suggest you really want to be using multiple projects. A keyword like `extends` makes sense to me and generally I would hope there is a single `environment.yml` which you can extend with some short lists of extra packages in the `project.yml`.

**Comment by mcg1969** [comment]
I agree that it _may_ be the case that orthogonal environments suggests multiple projects it need not be so. I think often we use the same environment for multiple streams simply out of convenience—because most development platforms effectively _dictate_ the use of a single environment.

I think we all agree that multiple environments has its uses, but forcing it to be "only" nested seems an unnecessary restriction. I don't see any point in being overly opinionated about it in the code. In fact, if I have a situation where I really don't have an obvious nesting, all I have to do is define a base environment that has nothing but Python in it (or no packages at all!) and "extend" my otherwise orthogonal environments around that.

**Comment by jbednar** [comment]
Good point; handling it as `extends` lets people do whatever they want while making it clear that it's normally expected that a project will have a single environment and some optional extra bits.

**Comment by maximlt** [comment]
I'm not sure that it's that uncommon to have multiple environments per project. `tox`, and others I guess, have made running commands in their own virtual environments very easy. I don't think that it bothers people so much.

I can easily imagine having a project with test dependencies and docs dependencies that are different enough so that I don't want to have to think about pinning versions so that I can get an environment that include both sets at the same time. We know too well how difficult this is do manage over time!

Basically I feel that if a project has more that one command and a test run command, it shouldn't really rely on the extending mechanism mentioned in this discussion.
